docs/guidance/head-command.md

```diff
@@ -0,0 +1,146 @@
+# Specify container commands for head Pod
+Users want to execute some commands at two timings:
+
+* (1) **Before `ray start`**: Take [this slack thread](https://ray-distributed.slack.com/archives/C02GFQ82JPM/p1675378764037199) as an example, the user wants to set up some environment variables that will be used by `ray start`.
+
+* (2) **After `ray start` (RayCluster is ready)**: Take [this slack thread](https://ray-distributed.slack.com/archives/C02GFQ82JPM/p1669647595429959) as an example, the user wants to launch a Ray serve deployment when the RayCluster is ready.
+
+## Current KubeRay operator behavior for container commands
+* The current behavior for container commands is very weird, and **may be updated in the future**.
+* See [code](https://github.com/ray-project/kuberay/blob/47148921c7d14813aea26a7974abda7cf22bbc52/ray-operator/controllers/ray/common/pod.go#L301-L326) for more details.
+
+## Timing 1: Before `ray start`
+Currently, for timing (1), we can set the container's `Command` and `Args` in RayCluster specification to reach the goal.
+
+```yaml
+# ray-operator/config/samples/ray-cluster.head-command.yaml
+    rayStartParams:
+        ...
+    #pod template
+    template:
+      spec:
+        containers:
+        - name: ray-head
+          image: rayproject/ray:2.2.0
+          resources:
+            ...
+          ports:
+            ...
+          # `command` and `args` will become a part of `spec.containers.0.args` in the head Pod.
+          command: ["echo 123"]
+```
+* Running head Pod
+  * `spec.containers.0.command` is hardcoded with `["/bin/bash", "-lc", "--"]`.
+  * `spec.containers.0.args` contains two parts:
+    * (Part 1) **user-specified command**: A string concatenates `headGroupSpec.template.spec.containers.0.command` from RayCluster and `headGroupSpec.template.spec.containers.0.args` from RayCluster together. In this example, the string will be `echo 123` because `headGroupSpec.template.spec.containers.0.args` is not defined here.
+    * (Part 2) **ray start command**: The command is created based on `rayStartParams` specified in RayCluster. The command will look like `ulimit -n 65536; ray start ...`.
+    * To summarize, `spec.containers.0.args` will be `$(user-specified command) && $(ray start command)`.
```

Comment:
I think this section is logically complete, but it still took some time for me to understand.  Is my understanding correct that the `spec.containers.0.command` and `spec.containers.0.args` provided by the user get modified and rewritten by the operator? That might be the confusing part, and it would help to start by stating that explicitly.  To be precise:

`spec.containers.0.args <-- spec.containers.0.command && ray start ...`
`spec.containers.0.command <-- ["/bin/bash", "-lc", "--"]`


Also, what happens if the user specifies `"args"`?

Reply:
My overall impression is that this modification logic is an implementation detail that most users won't care about, and we can remove it from the docs altogether (and maybe just point to the relevant part of the code for advanced users, or bump it below to a section called "Advanced")

My guess is that all users really want to know from this doc is that if you put `command: ["echo 123"]`, it will run before `ray start`. 

I could be wrong about this though, I'd be interested what other reviewers think.

Reply:
> Is my understanding correct that the `spec.containers.0.command` and `spec.containers.0.args` provided by the user get modified and rewritten by the operator?

Yes, users provide `headGroupSpec.template.spec.containers.0.command` and `headGroupSpec.template.spec.containers.0.args` in RayCluster CR, and they will be modified and rewritten by the operator to create `spec.containers.0.command` and `spec.containers.0.args`.

> Also, what happens if the user specifies "args"?

```
spec.containers.0.args <-- [headGroupSpec...command] [headGroupSpec...args] && [ray start ...]

# Example RayCluster
command: ["echo 123"]
args: ["456"]

# Pod args will be
echo 123  456  && [ray start ...]
```

Reply:
I added `args` to the RayCluster example. See https://github.com/ray-project/kuberay/pull/912/commits/f4a47ee702caf1f7c361e33c4f16885639c0486e for more details.